server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1029,6 +1029,9 @@ public CanMatchResponse canMatch(ShardSearchRequest request) throws IOException
             Rewriteable.rewrite(request.getRewriteable(), context, false);
             FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
             MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
+            if (indexShard.isSearchIdle() && searcher.getDirectoryReader().isCurrent() == false) {
```

Comment:
I don't understand the logic here. We call `acquireSearcherNoWrap` a few lines above, which calls `markSearcherAccessed` (i.e. setting `lastSearcherAccess` to currentTimeMillis). Then we check `indexShard.isSearchIdle()` here just a few milliseconds later, i.e., whether 30 seconds have passed since we last acquired the searcher? Are we not just always skipping this if clause here?

Reply:
oh you're right, the call to indexShard.isSearchIdle() should happen before we acquire the searcher. I thought that `markSearcherAccessed` was called only in `awaitShardSearchActive`.

Reply:
Also note that, if possible, I would prefer reusing the `refreshNeeded` method on `Engine` (which is overridden in ReadOnlyEngine) instead of the `searcher.getDirectoryReader().isCurrent() == false` check. Perhaps we can expose a new method on IndexShard that makes use of this as well as the  `isSearchIdle` method.